### ISSUE TITLE
Enable installation via cargo

### ISSUE THREAD

**Issue by charliermarsh** [charliermarsh/ruff#43]
We can't upload to crates.io because we're using an unreleased version of RustPython.

**Comment by charliermarsh** [comment]
I've asked about publishing Ruff's RustPython fork here: https://github.com/RustPython/RustPython/issues/4179

**Comment by youknowone** [comment]
Regardless we are ready or not, I want to distribute it before ruff 0.1. Please poke me when you are ready 0.1.

**Comment by charliermarsh** [comment]
@youknowone - Thank you, that's really helpful and appreciated.

There are a few PRs I'd need to merge to move off of my `RustPython` fork (which is largely ahead of `main`) -- I know you know about some of these, but I'll just enumerate them here for completeness:

- https://github.com/RustPython/RustPython/pull/4192
- https://github.com/RustPython/RustPython/pull/4143
- (Not totally essential) https://github.com/charliermarsh/RustPython/pull/3
- https://github.com/RustPython/RustPython/pull/4218

**Comment by charliermarsh** [comment]
RustPython just cut a new release on [crates.io](https://crates.io/crates/rustpython-parser). So the only blocker here is (1) confirming that we want to use the Crate, which might mean we're a little slower to pick up changes from Git; and (2) LibCST is still using my unpublished fork (which just removes pyo3 stuff).

**Comment by madkinsz** [comment]
Hey! I was looking at adding documentation for installation from crates.io then realized it wasn't supported :D

1) Is it feasible to use RustPython from Git for other releases and only publish to `crates.io` if the published RustPython version has the sufficient changes? Is that worth the trouble?

2) Is this resolved now?

**Comment by charliermarsh** [comment]
I think there are three blockers here that we'd need to resolve, since, in order to publish to `crates.io`, we can't have any Git dependencies in the dependency tree:

1. Publish our parser fork as their own crates (https://github.com/astral-sh/RustPython-Parser), e.g., `ruff_rustpython_parser` or similar.
2. In RustPython (and our fork), remove the Git dependency on [unicode_names2](https://github.com/astral-sh/RustPython-Parser/blob/7a3eedbf6fb4ea7068a1bf7fe0e97e963ea95ffd/Cargo.toml#LL41C1-L41C15).
3. Publish our LibCST fork as its own crate (https://github.com/charliermarsh/LibCST/tree/charlie/ruff).

**Comment by youknowone** [comment]
I'd like to grant publish permission of parser to anyone who need it.
I didn't finish unicode_names2 patch to merge it yet. I need to port the implementation to its own phf.

**Comment by T-256** [comment]
I think all mentioned blockers are gone, Are there any new?

**Comment by zanieb** [comment]
Yes now that we've resolved the blockers we are considering publishing to crates.io but have not come to a decision yet.

**Comment by KGrewal1** [comment]
Has there been any update on this (or would cargo install --git  https://github.com/astral-sh/ruff ruff work: seems to install but I'm not sure if there's tests I can run to confim)

**Comment by charliermarsh** [comment]
@KGrewal1 -- `cargo install --git  https://github.com/astral-sh/ruff ruff` should work just fine! I tested it myself too.

**Comment by code-yeongyu** [comment]
.

**Comment by philwo** [comment]
We're currently looking into options for a new Python code formatter for Chromium, and while we haven't made any decision yet, Ruff is definitely a favorite. 😊 

This issue is a blocker for us, though, as we have a policy that all third-party software has to be mirrored into our internal repository, and tools then have to be built from there using internal infrastructure for security reasons. This would be very easy to setup for me if Ruff and all of its dependencies were published to crates.io, as we have existing automation and processes for importing packages hosted there.

Would you be open to revisit this issue and see if Ruff could be published to crates.io?

**Comment by charliermarsh** [comment]
Thanks @philwo -- that's cool to hear! We can definitely consider it -- will need to discuss as a team. I assume it's not sufficient to, like, mirror the source distribution that we publish to PyPI, and build wheels from that?

**Comment by philwo** [comment]
Thank you @charliermarsh! I had a look - building Ruff locally on my developer machine from these sources works perfectly. But unfortunately the production environment, where I'd have to build and deploy the Ruff binary in the end, is very much optimized for the case where everything can be downloaded automatically from crates.io, and straying from that path quickly turns into a total headache. :/

In case that you would be fine in general with publishing Ruff's code to crates.io and making it buildable from there, and that there are only technical blockers and "someone has to do the work" remaining, would that be something we could help with by contributing?

**Comment by zanieb** [comment]
We sometimes release from forks of dependencies to get bug fixes to our users faster. I'm a little hesitant to lose the ability to do so. Of course, that idea is generally in conflict with downstream re-distributors.

I think the other concern is that people may consider our Rust libraries as part of our public API when, in reality, most (if not all) of our Rust APIs are unstable.

I'll make sure we talk about this as a team next week. It'd be great to see Ruff used for Chromium.

**Comment by BurntSushi** [comment]
One of the other issues here is that publishing `ruff` on crates.io, AFAIK, requires that _all_ of its dependencies are on crates.io too. `ruff` is split into a ton of different crates. So all of those would need to be published.

Today, when we make changes across crates (which is I think very common), we just make the changes. There's really nothing else to be done.

But if all of these crates are published to crates.io, then I think there is a fair bit of overhead here:

* The versions on crates that have been changed need to be bumped.
* The versions on crates depending on those crates need to have the dependency specification changed to reflect the new minimum version.
* The above two steps need to be repeated recursively up to `ruff` itself.

There is also the question of semver, but I think we could essentially bypass this entirely by using `0.0.x` versioning exclusively.

So I think that while the initial work to get everything published to crates.io is somewhat straight-forward, I think the ongoing burden of dealing with a bunch of independently published crates is pretty annoying.

My suspicion is that it's possible to automate the steps above. If that can be done in a way where our core development workflow can remain blissfully unaware of the fact that every internal crate is actually versioned, then I think that would mitigate my primary concern.

Is there tooling that you're aware of that helps with this problem?

**Comment by KGrewal1** [comment]
I believe it is possible to have `{workspace=true}` for crates within the workspace in which case the all the crates which are for internal use, could use that and only the version number in the workspace cargo.toml needs to be changed from 0.0.x to 0.0.x+1 (or however versioning internal crates is decided)

**Comment by MichaReiser** [comment]
@philwo just for my understanding. Are you limited to rust tooling (because they can only be installed from crates.io) or can you use other packages hosted on pypi (e.g. in case of black or any other Python tool)?

Also for my curiosity. For which Chromium project are you evaluating ruff?

**Comment by philwo** [comment]
@BurntSushi @KGrewal1 I don't have much experience with Rust yet, but this reminds me very much of the tradeoff of working in a monorepo vs. having to coordinate changes across multiple repos that depend on each other. The ability to make atomic changes easily across all crates is worth a lot, for sure! If `{workspace=true}` and `cargo smart-release` would allow you to continue doing that in your "monorepo" here, while also making it possible to easily publish stable release versions of Ruff to crates.io as a collection of individual, versioned crates that depend on each other, that seems like a super nice solution to me :)

@MichaReiser No, we're not limited to Rust tooling. We can import third-party open-source code written in most of the major languages into our monorepo and build it there. The problem is more that this import process tends to only support the canonical way how stuff is distributed in the respective language ecosystem, and doesn't handle special cases well.

For example, if I wanted to import a Go library that didn't use `go.mod` to manage its dependencies, but pulled them in via Git submodules or relied on a `vendor` folder managed via some custom script, it would likely not work.

Re: Black - we have an internal fork of Black called [Pyink](https://github.com/google/pyink) that stays in close sync with upstream and for all that matters to Chromium can be assumed to behave identically to upstream Black. Pyink's code is bi-di synced from and to GitHub on each change using automated tooling. Using Pyink / Black is the other option currently part of the proposal.

> Also for my curiosity. For which Chromium project are you evaluating ruff?

I'm part of a group of Chromium developers currently writing a proposal to change [Chromium's Python style guide](https://chromium.googlesource.com/chromium/src/+/main/styleguide/python/python.md) from YAPF to a more modern alternative. A few developers mentioned Ruff and that it has a couple of great benefits over Black, which is why I'm trying to find a way how to make it work, so that we can put it into the proposal as well. The change, if accepted by the Chromium Python developers committee, would then cover all of the Python code in Chromium's ecosystem, except ChromiumOS, which has [its own style guide](https://www.chromium.org/chromium-os/developer-library/reference/style-guides/python/).

**Comment by MichaReiser** [comment]
Thanks for the explanation. I assume something like `uv pip install ruff --no-binary :all:` that forces ruff to be built from the wheel's source wouldn't work for you.

**Comment by philwo** [comment]
> Thanks for the explanation. I assume something like `uv pip install ruff --no-binary :all:` that forces ruff to be built from the wheel's source wouldn't work for you.

Unfortunately not :( As we'd need to run the resulting `ruff` binary also on production servers, so that we could e.g. execute it to check formatting and possibly lint from Chromium's Gerrit or a web-based code editor, the security and auditing requirements for the builds and everything that goes into the resulting artifacts are really tough.

**Comment by CreatedBySeb** [comment]
Chiming in with a related need, at the company I work for we are able to build Rust-based packages from source and consume them in our software without them being published on crates.io, however those packages cannot have git dependencies, only local and crates.io dependencies, which I understand to also be a blocker for publishing on crates.io. I thought I would bring this up here, since it gives a different use case for the first portion of the work, which is removing any git dependencies from Ruff.

To provide context on the why, my company does third party library auditing when importing libraries. There is automated support for doing this from sources like crates.io and PyPI, and you can manually pull in the git source and do it yourself, however having direct git dependencies in `Cargo.toml` doesn't really fit in either box and is explicitly disallowed. Working around this would probably require build-time patching to redirect the dependencies to manually imported ones, which would get messy fast and slow adoption of new versions. It would be easiest for us to adopt new versions if all of Ruff's dependencies were available on crates.io, even if Ruff itself wasn't.

Currently the latest version of Ruff we have available internally is `v0.4.4`, since it doesn't have any git dependencies (`v0.4.5` introduced the `lsp-types` fork, and `v0.4.9` introduced salsa via git). There is significant appetite for using Ruff in my company (some teams are continuing to use the old version rather than go back to Flake8, black & isort), but we are missing out on lots of valuable improvements as a result of this issue which hampers widespread adoption. My team was recently re-evaluating Ruff and was interested in adopting it across our codebases, but the inability to adopt new versions is a blocker for us.

What are the possible paths forward for this? My understanding would be the following:

1. For `lsp-types`, either:
    a. Publish the fork, e.g. as `ruff-lsp-types`
    b. Upstream the changes to `lsp-types`, and swap the dependency to their published crate once they are released
2. For `salsa`, have them publish the changes in their master which are not yet released, and swap the dependency to the published crate

Publishing a fork for `salsa` also is technically an option, but since the issue isn't the changes being merged just them being released, it doesn't seem worthwhile. 

Happy to provide support if I can on unblocking this, since it will be significantly beneficial to my team and company more broadly.

**Comment by CreatedBySeb** [comment]
@BurntSushi Sure thing, essentially how the process normally works (taking Rust as an example) is that packages are imported from crates.io by a system that does supply chain checks for license, security, etc. They are then made available in an internal registry which cargo is directed to use, so when I build a package with cargo it can only access the vetted packages and anything else will not be found. Git dependencies bypass the registry entirely and so they are blocked by our build fleet since the packages are not vetted, which means that we cannot publish/consume packages with git dependencies. Similar processes are used for other languages with implementation differences based on that language's tooling, some of which facilitate patching more easily than others. This isn't a policy we are likely to be able to change, but it does also make reasonable sense since it would be hard to replicate this type of supply chain validation for Git dependencies.

**Comment by cjames23** [comment]
I just want to add one other wrinkle that we have, I work at the same company as Seb, locally our users this can work because they have network access to build but on our build system we are network jailed and in the case of packages like ruff we compile from source.

**Comment by chrisjsewell** [comment]
I would note rustpython now uses the ruff parser: https://github.com/RustPython/RustPython/pull/5494
It currently has to source from GitHub, which seems really not great at all 😅

**Comment by CreatedBySeb** [comment]
Since salsa has just done a release to `0.20.0` (https://github.com/salsa-rs/salsa/pull/753), could this be a good time to swap to a crates dependency on it rather than the git dependency? This release has the commit currently depended upon within it, so this shouldn't result in a regression for Ruff, but would remove one of the two git dependencies that are a blocker.

**Comment by CreatedBySeb** [comment]
In order to help minimise the effort required by the maintainers, I have made the minimal PR to change the versions having run the test suites on two devices to ensure it still works as expected: https://github.com/astral-sh/ruff/pull/17566

**Comment by MichaReiser** [comment]
I think we can now switch to using salsa as a normal dependency. But we first have to fix the release workflow upstream https://salsa.zulipchat.com/#narrow/channel/333573-Contributing-to-Salsa/topic/Release.200.2E26.2E1/with/580668398

**Comment by dhruvmanila** [comment]
This should be now possible (_after publishing our crates_) with the last git dependency being switched to a crate in https://github.com/astral-sh/ruff/pull/24894.

**Comment by MichaReiser** [comment]
Installing using cargo requires publishing our crates. But all git dependencies are now removed, so that we could publish our crates if we wanted so

**Comment by samuelcolvin** [comment]
as per https://github.com/pydantic/monty/issues/148 it would be extremely helpful for Pydantic Monty if ruff crates were uploaded to crates.io.

We use:
* `ruff_python_parser`
* `ruff_python_stdlib`
* `ruff_python_ast`
* `ruff_text_size`
* `ruff_db`
* `ty_python_semantic`
* `ty_python_core`
* `ty_module_resolver`
* `ty_vendored`

**Edit:** I just saw https://github.com/pydantic/monty/issues/148#issuecomment-4785168788 thank you @charliermarsh!

**Comment by samuelcolvin** [comment]
I'm afraid we still need the `ty_` crates before we can use this.

**Comment by charliermarsh** [comment]
Should all be out now.

**Comment by samuelcolvin** [comment]
That's great - thank you. We've adopted the published crates in Monty.

By the way, the package versions don't all match - this is what claude figured out for us:

```toml
ruff_python_parser = "0.0.3"
ruff_python_stdlib = "0.0.3"
ruff_python_ast = "0.0.3"
ruff_text_size = "0.0.3"
ruff_db = { version="0.0.3" ,features = ["serde"] }
# ty_python_semantic, ty_python_core and ty_vendored are only published at 0.0.2,
# but those 0.0.2 crates depend on the 0.0.3 line of the ruff crates and on
# ty_module_resolver 0.0.3, so the lower-level pins must be 0.0.3 to avoid two
# incompatible copies of each crate ending up in the dependency tree.
ty_python_semantic = "0.0.2"
ty_python_core = "0.0.2"
ty_module_resolver = "0.0.3"
ty_vendored = "0.0.2"
```

**Comment by charliermarsh** [comment]
Yeah because they got published at different increments. I'll probably adjust it though.
